File: content/snippets/python/design_patterns/state.py
from abc import ABC, abstractmethod

class State(ABC):
    @abstractmethod
    def handle(self) -> None:
        pass

class StartState(State):
    def handle(self) -> None:
        print("Entering start state")

class RunningState(State):
    def handle(self) -> None:
        print("Running state active")

class StoppedState(State):
    def handle(self) -> None:
        print("Stopped state active")

class Context:
    def __init__(self):
        self.state: State = None
    
    def set_state(self, state: State) -> None:
        self.state = state
    
    def request(self) -> None:
        self.state.handle()

ctx = Context()

ctx.set_state(StartState())
ctx.request()

ctx.set_state(RunningState())
ctx.request()

ctx.set_state(StoppedState())
ctx.request()

File: content/snippets/typescript/design_patterns/state.ts
interface IState {
    handle(): void;
}

class StartState implements IState {
    handle(): void {
        console.log("Entering start state");
    }
}

class RunningState implements IState {
    handle(): void {
        console.log("Running state active");
    }
}

class StoppedState implements IState {
    handle(): void {
        console.log("Stopped state active");
    }
}

class Context {
    private state: IState;
    
    setState(state: IState): void {
        this.state = state;
    }
    
    request(): void {
        this.state.handle();
    }
}

const ctx = new Context();

ctx.setState(new StartState());
ctx.request();

ctx.setState(new RunningState());
ctx.request();

ctx.setState(new StoppedState());
ctx.request();
